torch/random.py

```diff
@@ -0,0 +1,107 @@
+import torch
+import contextlib
+import warnings
+
+from torch._C import default_generator
+
+
+def set_rng_state(new_state):
+    r"""Sets the random number generator state.
+
+    Args:
+        new_state (torch.ByteTensor): The desired state
+    """
+    default_generator.set_state(new_state)
+
+
+def get_rng_state():
+    r"""Returns the random number generator state as a ByteTensor."""
+    return default_generator.get_state()
+
+
+def manual_seed(seed):
+    r"""Sets the seed for generating random numbers. And returns a
+    `torch._C.Generator` object.
+
+    Args:
+        seed (int or long): The desired seed.
+    """
+    import torch.cuda
+
+    if not torch.cuda._in_bad_fork:
+        torch.cuda.manual_seed_all(seed)
+
+    return default_generator.manual_seed(seed)
+
+
+def initial_seed():
+    r"""Returns the initial seed for generating random numbers as a
+    python `long`.
+    """
+    return default_generator.initial_seed()
+
+
+_fork_rng_warned_already = False
+
+@contextlib.contextmanager
+def fork_rng(devices=None, enabled=True, _caller="fork_rng", _devices_kw="devices"):
+    """
+    Forks the RNG, so that when you return, the RNG is reset
+    to the state that it was previously in.
+
+    Arguments:
+        devices (iterable of CUDA IDs): CUDA devices for which to fork
+            the RNG.  CPU RNG state is always forked.  By default, fork_rng operates
+            on all devices, but will emit a warning if your machine has a lot
+            of machines, since this function will run very slowly in that case.
+        enabled (bool): if False, the RNG is not forked.  This is a convenience
+            argument for easily disabling the context manager without having
+            to reindent your Python code.
+    """
+
+    import torch.cuda
+    global _fork_rng_warned_already
+
+    # Internal arguments:
+    #   _caller: the function which called fork_rng, which the user used
+    #   _devices_kw: the devices keyword of _caller
+
+    if not enabled:
+        yield
+        return
+
+    if devices is None or devices is True:
+        suppress_warning = devices is True
```

Comment:
what? why would devices be `True` (by identity!)? I thought it's a list of ids? If you want to test for emptiness you can't use the `is` operator

Reply:
OK, here's the back story: the code here always emits a warning when you have >1 GPU on your machine. So, what if you really did want all of the GPUs to be enabled? You might want to suppress the warning, in this case. `devices=None` has the correct semantics, but it emits warnings. I didn't want to force users to list them all out in the `devices` argument just to suppress the warning. So I made a new mode of calling (documented in the warning message), `devices=True`, which has the same semantics as `devices=None` but suppresses the warning.

We can just delete this logic and make the warning unsuppressable without specifying all devices manually if you like.

Reply:
This is quite awkward. Once someone learns about the pitfalls of this function, they will always have to write `devices=True`, which is not an acceptable for any other function in our framework that takes a `devices` argument

Reply:
OK the `devices=True` path has been axed.